contracts/vault-managers/CDPManager01.sol

```diff
@@ -0,0 +1,355 @@
+     * @param asset The address of the collateral token of a position
+     * @param user The owner of the position
+     **/
+    function triggerLiquidation(address asset, address user) external nonReentrant {
+
+        // check oracle
+        require(oracleRegistry.oracleByAsset(asset) != address(0), "Unit Protocol: DISABLED_ORACLE");
+
+        // USD value of the collateral
+        uint usdValue_q112 = IOracleUsd(oracleRegistry.oracleByAsset(asset)).assetToUsd(asset, vault.collaterals(asset, user));
+        
+        // reverts if a position is not liquidatable
+        require(_isLiquidatablePosition(asset, user, usdValue_q112), "Unit Protocol: SAFE_POSITION");
+
+        uint liquidationDiscount_q112 = usdValue_q112.mul(
+            vaultManagerParameters.liquidationDiscount(asset)
+        ).div(DENOMINATOR_1E5);
+
+        uint initialLiquidationPrice = usdValue_q112.sub(liquidationDiscount_q112).div(Q112);
+
+        // sends liquidation command to the Vault
+        vault.triggerLiquidation(asset, user, initialLiquidationPrice);
+
+        // fire an liquidation event
+        emit LiquidationTriggered(asset, user);
+    }
+
+    /**
+     * @dev Determines whether a position is liquidatable
+     * @param asset The address of the collateral
+     * @param user The owner of the position
+     * @param usdValue_q112 Q112-encoded USD value of the collateral
+     * @return boolean value, whether a position is liquidatable
+     **/
+    function _isLiquidatablePosition(
+        address asset,
+        address user,
+        uint usdValue_q112
+    ) internal view returns (bool) {
+        uint debt = vault.getTotalDebt(asset, user);
+
+        // position is collateralized if there is no debt
+        if (debt == 0) return false;
+
+        return debt.mul(100).mul(Q112).div(usdValue_q112) >= vaultManagerParameters.liquidationRatio(asset);
+    }
+
+    /**
+     * @dev Determines whether a position is liquidatable
+     * @param asset The address of the collateral
+     * @param user The owner of the position
+     * @return boolean value, whether a position is liquidatable
+     **/
+    function isLiquidatablePosition(
+        address asset,
+        address user
+    ) external view returns (bool) {
+        return utilizationRatio(asset, user) >= vaultManagerParameters.liquidationRatio(asset);
+    }
+
+    /**
+     * @dev Calculates current utilization ratio
+     * @param asset The address of the collateral
+     * @param user The owner of the position
+     * @return utilization ratio
+     **/
+    function utilizationRatio(
+        address asset,
+        address user
+    ) public view returns (uint) {
+        uint debt = vault.getTotalDebt(asset, user);
+        if (debt == 0) return uint(-1);
+        
+        uint usdValue_q112 = IOracleUsd(oracleRegistry.oracleByAsset(asset)).assetToUsd(asset, vault.collaterals(asset, user));
+
+        return debt.mul(100).mul(Q112).div(usdValue_q112);
+    }
```

Comment:
Copy-paste should be avoided.
`utilizationRatio` of zero debt is zero, isn't it?